File: frontend/src/services/cache.js
/**
 * Frontend cache service for API responses
 * Stores data in memory to avoid refetching on tab switches
 */

class FrontendCache {
    constructor() {
        this.cache = new Map();
        this.timestamps = new Map();
    }

    /**
     * Get cached data if still fresh
     * @param {string} key - Cache key
     * @param {number} ttl - Time to live in seconds (default: 60s)
     * @returns {any|null} - Cached data or null if expired/missing
     */
    get(key, ttl = 60) {
        if (!this.cache.has(key)) {
            return null;
        }

        const timestamp = this.timestamps.get(key);
        const age = (Date.now() - timestamp) / 1000; // age in seconds

        if (age > ttl) {
            // Expired - remove it
            this.cache.delete(key);
            this.timestamps.delete(key);
            console.log(`⏰ Cache expired: ${key}`);
            return null;
        }

        console.log(`✅ Cache hit: ${key} (age: ${age.toFixed(1)}s)`);
        return this.cache.get(key);
    }

    /**
     * Set cache data
     * @param {string} key - Cache key
     * @param {any} data - Data to cache
     */
    set(key, data) {
        this.cache.set(key, data);
        this.timestamps.set(key, Date.now());
        console.log(`💾 Cached: ${key}`);
    }

    /**
     * Invalidate specific key or all keys matching pattern
     * @param {string} keyOrPattern - Key or pattern to invalidate
     */
    invalidate(keyOrPattern) {
        if (keyOrPattern.endsWith('*')) {
            // Pattern match - remove all keys starting with pattern
            const prefix = keyOrPattern.slice(0, -1);
            const keysToDelete = Array.from(this.cache.keys()).filter(k => k.startsWith(prefix));
            keysToDelete.forEach(k => {
                this.cache.delete(k);
                this.timestamps.delete(k);
            });
            console.log(`🗑️ Invalidated: ${keyOrPattern} (${keysToDelete.length} items)`);
        } else {
            // Exact key
            this.cache.delete(keyOrPattern);
            this.timestamps.delete(keyOrPattern);
            console.log(`🗑️ Invalidated: ${keyOrPattern}`);
        }
    }

    /**
     * Clear all cache
     */
    clear() {
        this.cache.clear();
        this.timestamps.clear();
        console.log('🗑️ Cache cleared');
    }

    /**
     * Get cache stats
     */
    stats() {
        return {
            size: this.cache.size,
            keys: Array.from(this.cache.keys())
        };
    }
}

// Export singleton instance
export const frontendCache = new FrontendCache();

File: backend/app/utils/cache.py
"""
Simple in-memory cache for API responses with TTL (Time To Live)
"""
from datetime import datetime, timedelta
from typing import Any, Optional
import hashlib
import json

class SimpleCache:
    def __init__(self):
        self._cache = {}
    
    def _make_key(self, endpoint: str, params: dict) -> str:
        """Create cache key from endpoint and params"""
        param_str = json.dumps(params, sort_keys=True)
        return f"{endpoint}:{hashlib.md5(param_str.encode()).hexdigest()}"
    
    def get(self, endpoint: str, params: dict = None) -> Optional[Any]:
        """Get cached value if not expired"""
        params = params or {}
        key = self._make_key(endpoint, params)
        
        if key in self._cache:
            value, expiry = self._cache[key]
            if datetime.now() < expiry:
                print(f"✅ Cache HIT: {endpoint}")
                return value
            else:
                # Expired - remove it
                del self._cache[key]
                print(f"⏰ Cache EXPIRED: {endpoint}")
        
        print(f"❌ Cache MISS: {endpoint}")
        return None
    
    def set(self, endpoint: str, value: Any, ttl_seconds: int = 60, params: dict = None):
        """Set cache value with TTL"""
        params = params or {}
        key = self._make_key(endpoint, params)
        expiry = datetime.now() + timedelta(seconds=ttl_seconds)
        self._cache[key] = (value, expiry)
        print(f"💾 Cache SET: {endpoint} (TTL: {ttl_seconds}s)")
    
    def invalidate(self, endpoint: str = None):
        """Clear cache for specific endpoint or all"""
        if endpoint:
            # Clear all keys starting with endpoint
            keys_to_remove = [k for k in self._cache.keys() if k.startswith(endpoint + ":")]
            for key in keys_to_remove:
                del self._cache[key]
            print(f"🗑️  Cache INVALIDATED: {endpoint}")
        else:
            self._cache.clear()
            print("🗑️  Cache CLEARED (all)")
    
    def size(self) -> int:
        """Get number of cached items"""
        return len(self._cache)

# Global cache instance
api_cache = SimpleCache()
